sum of last digits of two given numbers.last digit of 267 is 7.last digit of 154 is 4.sum of 7 and 4 =11.

In [ ]:
class sum_of_last_digits:
  def __init__(self,n1,n2):
    self.n1=n1
    self.n2=n2
  def last_digit(self,n):
    return abs(n)%10
  def sum_of_digits(self,n1,n2):
    return self.last_digit(self.n1)+self.last_digit(self.n2)
n1=264 #264
n2=-157 #-157
res=sum_of_last_digits(n1,n2)
print("sum:",res.sum_of_digits(n1,n2))

sum: 11


In [ ]:
def last_digit(n):
  return abs(n)%10
def sum_of_digits(n1,n2):
  return last_digit(n1)+last_digit(n2)
n1=int(input())
n2=int(input())
res=sum_of_digits(n1,n2)
print("sum:",res)

124
1567
sum: 11


In [ ]:
def last_digit(n):
  return n%10
def sum_of_last_digit(n1,n2):
  return last_digit(n1)+last_digit(n2)
n1=int(input())
n2=int(input())
res=sum_of_last_digit(n1,n2)
print("sum:",res)

34
45
sum: 9


In [ ]:
def last_digit(n):
  if n<0:
    n=-n
  return n%10
def sum_of_last_digit(n1,n2):
  return last_digit(n1)+last_digit(n2)
n1=int(input())
n2=int(input())
res=sum_of_last_digit(n1,n2)
print("sum:",res)

34
-78
sum: 12


Write a program to find the sum of Squares
Input : N = 4
Output : 30
1^2 + 2^2 + 3^2 + 4^2
= 1 + 4 + 9 + 16
= 30
Input : N = 5
Output : 55

In [ ]:
class sum_of_squares:
  def __init__(self,n):
    self.n=n
  def squaring(self):
    sum=0
    for i in range(1,self.n+1):
      sum+=i**2
    return sum
n=int(input())
res=sum_of_squares(n)
print("sum of squares of ",n,"is",res.squaring())

5
sum of squares of  5 is 55


In [ ]:
def sum_of_squares(n):
  sum=0
  for i in range(1,1+n):
    sum+=i**2
  return sum
n=int(input())
result=sum_of_squares(n)
print("sum of squares:",result)

4
sum of squares: 30


Write a program to input amount from user and print minimumnumber of notes (Rs. 500, 100, 50, 20, 10, 5, 2, 1) required for the amount.
Example
Input
Input amount: 575
Output
Total number of notes:
500: 1
100: 0
50: 1
20: 1
10: 0
5: 1
2: 0
1: 0


In [ ]:
def num_of_notes(amt):
  notes=[2000,500,200,100,50,20,10,5,2,1]
  notes_dict={}
  for note in notes:
    notes_dict[note]=amt//note
    amt=amt%note
  return notes_dict

amt=int(input())
res=num_of_notes(amt)
for note,count in res.items():
  print(f"{note}:{count}")

2567
2000:1
500:1
200:0
100:0
50:1
20:0
10:1
5:1
2:1
1:0


In [ ]:
class notes_calculation:
  def __init__(self,amt):
    self.amt=amt
  def num_of_notes(self):
    notes=[2000,500,200,100,50,20,10,5,2,1]
    notes_dict={}
    for i in notes:
      notes_dict[i]=self.amt//i
      self.amt=self.amt%i
    return notes_dict
amt=int(input())
result=notes_calculation(amt)
res=result.num_of_notes()
for i,count in res.items():
  print(f"{i}:{count}")

546
2000:0
500:1
200:0
100:0
50:0
20:2
10:0
5:1
2:0
1:1


Inverted pyramid pattern

In [ ]:
n=5
for i in range(n,0,-1):
  print(" ","*"*i)


  *****
  ****
  ***
  **
  *


In [ ]:
n=5
for i in range(n,0,-1):
  print(" "*(n-i)+"* "*i)

* * * * * 
 * * * * 
  * * * 
   * * 
    * 


In [ ]:
n=5
for i in range(n,0,-1):
  print(" "*(n-i)+"*"*i)

*****
 ****
  ***
   **
    *


In [ ]:
def inverted_pyramid_pattern(rows):
  for i in range(rows,0,-1):
    print(" "*(rows-i)+"* "*i)

rows=int(input())
result=inverted_pyramid_pattern(rows)




6
* * * * * * 
 * * * * * 
  * * * * 
   * * * 
    * * 
     * 


In [ ]:
class inverted_pattern:
  def __init__(self,rows):
    self.rows=rows
  def pattern(self):
    for i in range(self.rows,0,-1):
      print(" "*(self.rows-i)+"* "*i)
n=int(input("enter num of rows:"))
res=inverted_pattern(n)
res.pattern()

enter num of rows:5
* * * * * 
 * * * * 
  * * * 
   * * 
    * 


pattern

```
1
3 2
6 5 4
10 9 8 7
```



In [ ]:
rows=int(input("enter number of rows:"))
start=1
for i in range(1,rows+1):
  end=start+i-1
  for j in range(end,start-1,-1):
    print(j,end=" ")
  print()
  start=end+1


enter number of rows:4
1 
3 2 
6 5 4 
10 9 8 7 


In [ ]:
class pattern:
  def __init__(self,rows):
    self.rows=rows
  def display(self):
    start=1
    for i in range(1,self.rows+1):
      end=start+i-1
      for j in range(end,start-1,-1):
        print(j,end=" ")
      print()
      start=end+1
result=pattern(5)
result.display()


1 
3 2 
6 5 4 
10 9 8 7 
15 14 13 12 11 


In [ ]:
class pattern:
  @classmethod
  def display(cls,rows):
    start=1
    for i in range(1,rows+1):
      end=start+i-1
      for j in range(end,start-1,-1):
        print(j,end=" ")
      print()
      start=end+1
pattern.display(4)

1 
3 2 
6 5 4 
10 9 8 7 


count the number of even numbers with 5 input

In [ ]:
class even(object):
  @classmethod
  def count_evens(cls,n1,n2,n3,n4,n5):
    num_list=[n1,n2,n3,n4,n5]
    for i in num_list:
      return sum(1 for i in num_list if abs(i)%2==0)
result=even.count_evens(12,17,-18,9,88)
print(result)

3


In [ ]:
class even:
  def __init__(self,nums):
    self.nums=nums
  def even_count(self):
    return sum(1 for i in self.nums if abs(i)%2==0)
nums=[int(input(f"enter the num {i+1}:"))for i in range(5)]
result=even(nums)
res=result.even_count()
print(res)

enter the num 1:12
enter the num 2:14
enter the num 3:18
enter the num 4:6
enter the num 5:7
4


In [ ]:
nums=[int(input(f"enter num {i+1}:"))for i in range(5)]
even_count=sum(1 for i in nums if abs(i)%2==0)
print("even count:",even_count)

enter num 1:-9
enter num 2:-64
enter num 3:78
enter num 4:90
enter num 5:-4
even count: 4


A circular prime is a prime number with the property that the number generated at each intermediate step when cyclically permuting its digits will be prime. For example, 1193 is a circular prime, since 1931, 9311 and 3119 all are also prime.

A circular prime is a prime number that remains prime even after any cyclic rotation of its digits.

Task Write a program that takes an integer N as input and determines whether it is a circular prime.

Input Format

A single integer N (1 ≤ N ≤ 10⁶)

Output Format

Print "Circular Prime" if N is a circular prime.

Otherwise, print "Not a Circular Prime".

Example Test Cases Input 1:

1193

Output 1:

Circular Prime

Input 2:

197

Output 2:

Circular Prime

Input 3:

23

Output 3:

Circular Prime

Input 4:

35

Output 4:

Not a Circular Prime

In [ ]:
def is_prime_check(n):
  if n<2:
    return False
  for i in range(2,int(n**0.5)+1):
    if n%i==0:
      return False
  return True
def rotations(n):
  str_n=str(n)
  return [int(str_n[i:]+str_n[:i])for i in range(len(str_n))]
def check(n):
  num_rotation_list=rotations(n)
  return all(is_prime_check(i) for i in num_rotation_list)
n=int(input())
if check(n):
  print("it is circular prime!")
else:
  print("it is not a circular prime!!")

197
it is circular prime!


In [ ]:
class circular_prime:
  @classmethod
  def is_prime(cls,n):
    if n<2:
      return False
    for i in range(2,int(n**0.5)+1):
      if n%i==0:
        return False
    return True
  @classmethod
  def rotations(cls,n):
    str_n=str(n)
    return [int(str_n[i:]+str_n[:i])for i in range(len(str_n))]
  @classmethod
  def check(cls,n):
    rotation_list=cls.rotations(n)
    return all(cls.is_prime(i) for i in rotation_list)
n=int(input())
if circular_prime.check(n):
  print("circular prime!!")
else:
  print("not a circular prime!!")



197
circular prime!!


In [ ]:
class circular_prime:
  def __init__(self,n):
    self.n=n
  def is_prime(self,n):
    if self.n<2:
      return False
    for i in range(2,int(self.n**0.5)+1):
      if self.n%i==0:
        return False
    return True
  def rotations(self):
    str_n=str(self.n)
    return [int(str_n[i:]+str_n[:i])for i in range(len(str_n))]
  def check(self):
    rotation_list=self.rotations()
    return all(self.is_prime(i) for i in rotation_list)
n=int(input())
result=circular_prime(n)
if result.check():
  print("circular prime!!")
else:
  print("not a circular prime!!")



1193
circular prime!!


prime number

In [ ]:
def is_prime_check(n):
  if n<2:
    return False
  for i in range(2,int(n**0.5)+1):
    if n%i==0:
      return False
  return True
n=int(input())
if(is_prime_check(n)):
  print("prime num!")
else:
  print("not a prime num!")


17
prime num!


In [ ]:
class prime_check:
  def __init__(self,n):
    self.n=n
  def check(self):
    if self.n<2:
      return False
    for i in range(2,(int(self.n**0.5)+1)):
      if self.n%i==0:
        return False
    return True
  def display(self):
    if self.check():
      print("prime!")
    else:
      print("not prime!")
n=int(input())
res=prime_check(n)
res.display()


14
not prime!


In [ ]:
class prime_check:
  @classmethod
  def check(cls,n):
    if n<2:
      return False
    for i in range(2,int(n**0.5)+1):
      if n%i==0:
        return False
    return True
  @classmethod
  def display(cls,n):
    if cls.check(n):
      print("prime!")
    else:
      print("not prime!")
n=int(input())
prime_check.display(n)

3
prime!


In [ ]:
def prime_check(n):
  if n<2:
    return False
  for i in range(2,n):
    if n%i==0:
      return False
  return True
n=int(input())
if(prime_check(n)):
  print("prime num!")
else:
  print("not a prime num!")

17
prime num!


In [ ]:
class prime_check:
  def __init__(self,n):
    self.n=n
  def check(self):
    if self.n<2:
      return False
    for i in range(2,self.n):
      if self.n%i==0:
        return False
    return True
  def display(self):
    if(self.check()):
      print("prime!")
    else:
      print("not prime!")
n=int(input())
res=prime_check(n)
res.display()

17
prime!


In [ ]:
class prime_check:
  @classmethod
  def check(cls,n):
    if n<2:
      return False
    for i in range(2,n):
      if n%i==0:
        return False
    return True
  @classmethod
  def display(cls,n):
    if(cls.check(n)):
      print("prime!")
    else:
      print("not a prime!")
n=int(input())
prime_check.display(n)

13
prime!


PRIME NUMBER IN RANGE

In [ ]:
def prime_check(n):
  if n<2:
    return False
  for i in range(2,n):
    if n%i==0:
      return False
  return True
def find_range(start,end):
  return [i for i in range(start,end) if prime_check(i)]
def display(start,end):
  prime_nums=find_range(start,end)
  if prime_nums:
    print(*prime_nums)
  else:
    print("no prime num in range")
start=int(input())
end=int(input())
display(start,end)

10
100
11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97


In [ ]:
class prime_range:
  def __init__(self,start,end):
    self.start=start
    self.end=end
  def check(self,n):
    if n<2:
      return False
    for i in range(2,n):
      if n%i==0:
        return False
    return True
  def find_range(self):
    return [i for i in range(self.start,self.end) if self.check(i)]
  def display(self):
    prime_num=self.find_range()
    if prime_num:
      print(*prime_num)
    else:
      print("no prime num in range!")
start=int(input())
end=int(input())
prime_range(start,end).display()

10
50
11 13 17 19 23 29 31 37 41 43 47


In [ ]:
class prime_range:
  @classmethod
  def check(cls,n):
    if n<2:
      return False
    for i in range(2,n):
      if n%i==0:
        return False
    return True
  @classmethod
  def find_range(cls,start,end):
    return [i for i in range(start,end) if cls.check(i)]
  @classmethod
  def display(cls,start,end):
    prime_num=cls.find_range(start,end)
    if prime_num:
      print(*prime_num)
    else:
      print("no prime num in range!")
start=int(input())
end=int(input())
prime_range.display(start,end)



10
20
11 13 17 19


Write a program that takes an integer N as input and converts its digits into words.

Input Format A single integer N (1 ≤ N ≤ 10⁶)

Output Format Print the digits of N in word format, separated by spaces.

Input 1:

Number=1

Output:

Digit=1

Words=One

Input 2:

Enter number=7777

Output:

Digit=7777

Words=Seven Seven Seven Seven

Input 3:

Enter number=1010

Output:

Digit=1010
Input 4:

Enter number=007

Output:

Digit=7

Words=Seven

In [ ]:
words={
    0:"Zero",
    1:"One",
    2:"Two",
    3:"Three",
    4:"Four",
    5:"Five",
    6:"Six",
    7:"Seven",
    8:"Eight",
    9:"Nine"
}
n=input("enter a num:").lstrip('0')
result=" ".join(words[int(i)] for i in n)
print(f"digit:{n}")
print(f"word:{result}")

enter a num:002548
digit:2548
word:Two Five Four Eight


In [ ]:
class num_words:
  def __init__(self,n):
    self.n=n
    self.words={
        0:"Zero",
    1:"One",
    2:"Two",
    3:"Three",
    4:"Four",
    5:"Five",
    6:"Six",
    7:"Seven",
    8:"Eight",
    9:"Nine"
  }
  def to_words(self):
    return " ".join(self.words[int(i)] for i in self.n)
  def display(self):
    print(f"digit:{self.n}")
    print(f"word:{self.to_words()}")
n=input().lstrip('0')
result=num_words(n)
result.display()

8970
digit:8970
word:Eight Nine Seven Zero


In [ ]:
class num_words:
  words={0:"Zero",
    1:"One",
    2:"Two",
    3:"Three",
    4:"Four",
    5:"Five",
    6:"Six",
    7:"Seven",
    8:"Eight",
    9:"Nine"
    }
  @classmethod
  def to_words(cls,n):
    res=" ".join(words[int(i)] for i in n)
    print(f"digit:{n}")
    print(f"word:{res}")
n=input().lstrip('0')
num_words.to_words(n)


908
digit:908
word:Nine Zero Eight


You are provided with 4 numbers: input1, input2, input3 and input4 and are expected to

find a secret PIN

The three numbers input1, input2 and input3 are four digit numbers within the range >=1000 and <=9999. ie.

1000 <= input1 <= 9999

1000 <= input2 <= 9999

1000 <= input3 <= 9999

input4 is a positive integer number.

PIN = ((smallest digit in input1) x (LARGEST digit in input2) x (smallest digit in input3)) - input

Example-if input1 = 3521, input2=2452, input3=1352, and input4-38, then PIN = (1 x 5x1)-38--33

Assuming that the 4 numbers are passed to the given function, Complete the function to find and retum the PIN....

In [ ]:
def small(n):
  min=9
  while n>0:
    digit=n%10
    if digit<min:
      min=digit
    n=n//10
  return min
def large(n):
  max=0
  while n>0:
    digit=n%10
    if digit>max:
      max=digit
    n=n//10
  return max
n1=int(input())
n2=int(input())
n3=int(input())
n4=int(input())
pin=small(n1)*large(n2)*small(n3)-n4
print(pin)


1233
3425
1234
38
-33


In [ ]:
class secret_pin:
  @classmethod
  def small(cls,n):
    min=9
    while n>0:
      digit=n%10
      if digit<min:
        min=digit
      n=n//10
    return min
  @classmethod
  def large(cls,n):
    max=0
    while n>0:
      digit=n%10
      if digit>max:
        max=digit
      n=n//10
    return max
  @classmethod
  def pin(cls,n1,n2,n3,n4):
    return cls.small(n1)*cls.large(n2)*cls.small(n3)-n4
n1=int(input())
n2=int(input())
n3=int(input())
n4=int(input())
result=secret_pin.pin(n1,n2,n3,n4)
print(result)



1254
1254
3512
37
-32


In [ ]:
class secret_pin:
  def __init__(self,n1,n2,n3,n4):
    self.n1=n1
    self.n2=n2
    self.n3=n3
    self.n4=n4
  def small(self,n):
    min=9
    while n>0:
      digit=n%10
      if digit<min:
        min=digit
      n=n//10
    return min
  def large(self,n):
    max=0
    while n>0:
      digit=n%10
      if digit>max:
        max=digit
      n=n//10
    return max
  def pin(self):
    return self.small(self.n1)*self.large(self.n2)*self.small(self.n3)-self.n4
n1=int(input())
n2=int(input())
n3=int(input())
n4=int(input())
res=secret_pin(n1,n2,n3,n4)
print(res.pin())



1234
1245
31247
38
-33


In [ ]:
def secret_pin(n1,n2,n3,n4):
  min_n1=min(str(n1))
  max_n2=max(str(n2))
  min_n3=min(str(n3))
  return (int(2min_n1)*int(max_n2)*int(min_n3)-n4)
n1=int(input())
n2=int(input())
n3=int(input())
n4=int(input())
result=secret_pin(n1,n2,n3,n4)
print(result)

1215
5415
2154
38
-33


In [ ]:
class secret_pin:
  def __init__(self,n1,n2,n3,n4):
    self.n1=n1
    self.n2=n2
    self.n3=n3
    self.n4=n4
  def pin(self):
    min_n1=min(str(self.n1))
    max_n2=max(str(self.n2))
    min_n3=min(str(self.n3))
    return (int(min_n1)*int(max_n2)*int(min_n3)-self.n4)
n1=int(input())
n2=int(input())
n3=int(input())
n4=int(input())
res=secret_pin(n1,n2,n3,n4)
print(res.pin())


124
3254
5124
38
-33


In [ ]:
class secret_pin:
  @classmethod
  def pin(cls,n1,n2,n3,n4):
    min_n1=min(str(n1))
    max_n2=max(str(n2))
    min_n3=min(str(n3))
    return (int(min_n1)*int(max_n2)*int(min_n3)-n4)
n1=int(input())
n2=int(input())
n3=int(input())
n4=int(input())
res=secret_pin.pin(n1,n2,n3,n4)
print(res)

1245
3254
1245
38
-33


In [ ]:
def small(n):
  return min(map(int,str(n)))
def large(n):
  return max(map(int,str(n)))
n1=int(input())
n2=int(input())
n3=int(input())
n4=int(input())
pin=small(n1)*large(n2)*small(n3)-n4
print(pin)

124
524
124
38
-33


In [ ]:
class secret_pin:
  @classmethod
  def small(cls,n):
    return min(map(int,str(n)))
  @classmethod
  def large(cls,n):
    return max(map(int,str(n)))
n1=int(input())
n2=int(input())
n3=int(input())
n4=int(input())
pin=secret_pin.small(n1)*secret_pin.large(n2)*secret_pin.small(n3)-n42
print(pin)




124
15
145
38
-33


In [ ]:
class secret_pin:
  def __init__(self,n1,n2,n3,n4):
    self.n1=n1
    self.n2=n2
    self.n3=n3
    self.n4=n4
  def small(self,n):
    return min(map(int,str(n)))
  def large(self,n):
    return max(map(int,str(n)))
  def pin(self):
    return self.small(self.n1)*self.large(self.n2)*self.small(self.n3)-self.n4
n1=int(input())
n2=int(input())
n3=int(input())
n4=int(input())
res=secret_pin(n1,n2,n3,n4)
print(res.pin())


124
3254
127
38
-33


max and min in given num

In [ ]:
def min_max(n):
  min=9
  max=0
  while n>0:
    digit=n%10
    if digit<min:
      min=digit
    if digit>max:
      max=digit
    n=n//10
  return min,max
n1=int(input())
min_digit,max_digit=min_max(n1)
print(min_digit)
print(max_digit)

1548
1
8


In [ ]:
class min_max:
  def __init__(self,n):
    self.n=n
    min=9
    max=0
    while self.n>0:
      digit=self.n%10
      if digit<min:
        min=digit
      if digit>max:
        max=digit
      self.n=self.n//10
    self.min_value=min
    self.max_value=max
n=int(input())
res=min_max(n)
print(res.min_value)
print(res.max_value)


1248
1
8


In [ ]:
class min_max:
  @classmethod
  def find(cls,n):
    min=9
    max=0
    while n>0:
      digit=n%10
      if digit<min:
        min=digit
      if digit>max:
        max=digit
      n=n//10
    return min,max
n=int(input())
min_digit,max_digit=min_max.find(n)
print(min_digit)
print(max_digit)


1246
1
6


Problem Statement:

A number is called strong number if sum of the factorial of its digit is equal to number itself.

For example: 145

since

1! + 4! + 5! = 1 + 24 + 120 = 145

So, 145 is a Strong number. Now given a positive number N and you have to find the number is strong number or not .

Input:

An integer T (1<=T<=1000) denoting the number of test cases followed by T lines. Each next line containing a single integer N (0 <= N<= 1018)

Output:

For each case output string “YES” if given number is strong number and “NO” otherwise in next line.

Sample Input
2
145
12
Sample Output
YES


In [ ]:
def factorial(n):
  if n==0 or n==1:
    return 1
  else:
    return n*factorial(n-1)
def sum_factorial(n):
  sum=0
  num=n
  while num>0:
    digit=num%10
    sum=sum+factorial(digit)
    num=num//10
  return sum
n=int(input())
fact=sum_factorial(n)
if n==fact:
  print("strong!")
else:
  print("not strong!")

145
strong!


In [ ]:
class strong_num:
  def __init__(self,n):
    self.n=n
  def factorial(self,n):
    if n==0 or n==1:
      return 1
    else:
      return n*self.factorial(n-1)
  def sum_fact(self):
    sum=0
    temp=self.n
    while temp>0:
      digit=temp%10
      sum=sum+self.factorial(digit)
      temp=temp//10
    return sum
  def check(self):
    if self.n==self.sum_fact():
      print("strong num!")
    else:
      print("not a strong num!")
n=int(input())
strong_num(n).check()

145
strong num!


In [ ]:
class strong_num:
  @classmethod
  def factorial(cls,n):
    if n==0 or n==1:
      return 1
    else:
      return n*cls.factorial(n-1)
  @classmethod
  def sum_fact(cls,n):
    sum=0
    temp=n
    while temp>0:
      digit=temp%10
      sum=sum+cls.factorial(digit)
      temp//=10
    return sum
  @classmethod
  def check(cls,n):
    if n==cls.sum_fact(n):
      print("strong num!")
    else:
      print("not strong num!")
n=int(input())
strong_num.check(n)

145
strong num!


In [ ]:
def strong_num(n):
  temp=n
  sum=0
  while n>0:
    digit=n%10
    fact=1
    for i in range(1,digit+1):
      fact=fact*i
    sum=sum+fact
    n=n//10
  if (temp==sum):
    print("it is strong_num!")
  else:
    print("not a strong num!")
n=int(input())
strong_num(n)




145
it is strong_num!


In [ ]:
class strong_num:
  def __init__(self,n):
    self.n=n
  def sum_fact(self):
    temp=self.n
    sum=0
    while temp>0:
      digit=temp%10
      fact=1
      for i in range(1,digit+1):
        fact=fact*i
      sum=sum+fact
      temp=temp//10
    return sum
  def check(self):
    if self.n==self.sum_fact():
      print("strong num!")
    else:
      print("not a strong num!")
n=int(input())
strong_num(n).check()



145
strong num!


factorial


In [ ]:
def factorial(n):
  if n==0 or n==1:
    return 1
  else:
    return n*factorial(n-1)
n=int(input())
result=factorial(n)
print(result)

3
6


In [ ]:
class factorial:
  def __init__(self,n):
    self.n=n
  def check(self):
    if self.n==0 or self.n==1:
      return 1
    else:
      return self.n*factorial(self.n-1).check()
n=int(input())
res=factorial(n)
print(res.check())

5
120


In [ ]:
class factorial:
  def __init__(self,n):
    self.n=n
  def check(self):
    if self.n==0 or self.n==1:
      return 1
    else:
      return self.n*factorial(self.n-1).check()
  def display(self):
    print(self.check())
n=int(input())
res=factorial(n)
res.display()

5
120


In [ ]:
class factorial:
  @classmethod
  def check(cls,n):
    if n==0 or n==1:
      return 1
    else:
      return n*cls.check(n-1)
n=int(input())
print(factorial.check(n))

5
120


In [ ]:
def factorial(n):
  fact=1
  for i in range(1,n+1):
    fact=fact*i
  return fact
n=int(input())
res=factorial(n)
print(res)


5
120


In [ ]:
class factorial:
  def __init__(self,n):
    self.n=n
  def check(self):
    fact=1
    for i in range(1,self.n+1):
      fact=fact*i
    return fact
  def display(self):
    print(self.check())
n=int(input())
factorial(n).display()

5
120


In [ ]:
class factorial:
  @classmethod
  def check(cls,n):
    fact=1
    for i in range(1,n+1):
      fact=fact*i
    return fact
  @classmethod
  def display(cls,n):
    print(cls.check(n))
n=int(input())
factorial().display(n)

5
120


In [ ]:
pp=0
qq=6
rr=7
pp=rr+pp
pp=(rr&4)^rr
if((qq&pp&rr)<(rr&qq)):
    if((qq^pp)<(rr+qq)):
        rr=(3+1)^pp
print(pp+qq+rr)


16


In [ ]:
pp=9
qq=6
rr=10
pp=rr+pp
pp=(rr&4)^rr
if((qq&pp&rr)<(rr&qq)):
    if((qq^pp)<(rr+qq)):
        rr=(3+1)^pp
print(pp+qq+rr)

26


maximum sum of sub array in array using sliding window


In [ ]:
def sub_arr_sum(arr,k):
  windows_sum=sum(arr[:k])
  max_sum=windows_sum
  for i in range(len(arr)-k):
    windows_sum=windows_sum-arr[i]+arr[i+k]
    max_sum=max(windows_sum,max_sum)
  return max_sum
arr=[2,1,5,1,3,2]
k=3
print(sub_arr_sum(arr,k))

9


In [ ]:
class sub_arr_sum:
  def __init__(self,arr,k):
    self.arr=arr
    self.k=k
  def find_max(self):
    windows_sum=sum(self.arr[:self.k])
    max_sum=windows_sum
    for i in range(len(self.arr)-self.k):
      windows_sum=windows_sum-self.arr[i]+self.arr[i+self.k]
      max_sum=max(windows_sum,max_sum)
    return max_sum
arr=[1,2,5,1,3,2]
k=3
print(sub_arr_sum(arr,k).find_max())


9


In [ ]:
class sub_arr_sum:
  @classmethod
  def find_max(cls,arr,k):
    windows_sum=sum(arr[:k])
    max_sum=windows_sum
    for i in range(len(arr)-k):
      windows_sum=windows_sum-arr[i]+arr[i+k]
      max_sum=max(windows_sum,max_sum)
    return max_sum
arr=[1,2,5,1,3,2]
k=3
print(sub_arr_sum().find_max(arr,k))

9


max_sub array in array using sliding window

In [ ]:
def max_sum_sub_arr(arr,k):
  windows_sum=sum(arr[:k])
  max_sum=windows_sum
  index=0
  for i in range(len(arr)-k):
    windows_sum=windows_sum-arr[i]+arr[i+k]
    if windows_sum>max_sum:
      max_sum=windows_sum
      index=i+1
  return max_sum,arr[index:index+k]
arr=[1,2,5,1,3,2]
k=3
max,subarray=max_sum_sub_arr(arr,k)
print(max)
print(subarray)

9
[5, 1, 3]


In [ ]:
class max_sum_sub_arr:
  def __init__(self,arr,k):
    self.arr=arr
    self.k=k
  def find(self):
    windows_sum=sum(self.arr[:self.k])
    max_sum=windows_sum
    index=0
    for i in range(len(self.arr)-self.k):
      windows_sum=windows_sum-self.arr[i]+self.arr[i+self.k]
      if windows_sum>max_sum:
        max_sum=windows_sum
        index=i+1
    return max_sum,self.arr[index:index+self.k]
arr=[1,2,5,1,3,2]
k=3
max,subarray=max_sum_sub_arr(arr,k).find()
print(max)
print(subarray)

9
[5, 1, 3]


In [ ]:
class max_sum_sub_arr:
  @classmethod
  def find(cls,arr,k):
    windows_sum=sum(arr[:k])
    max_sum=windows_sum
    index=0
    for i in range(len(arr)-k):
      windows_sum=windows_sum-arr[i]+arr[i+k]
      if windows_sum>max_sum:
        max_sum=windows_sum
        index=i+1
    return max_sum,arr[index:index+k]
arr=[1,2,5,1,3,2]
k=3
max,subarray=max_sum_sub_arr().find(arr,k)
print(max)
print(subarray)

9
[5, 1, 3]




```
i/p: arr=[12,3,14,56,77,13]
num=12
diff=2
o/p: 3
 condition:num-arr[i]<=diff
 then count++
 ```



In [ ]:
def find(arr,num,diff):
  count=0
  for i in range(len(arr)):
    if(abs(arr[i]-num)<=diff):
      count+=1
  return count
arr=list(map(int,input().split()))
num=int(input())
diff=int(input())
print(find(arr,num,diff))



12 3 14 56 77 13
13 
2
3


In [ ]:
class comb_count:
  def __init__(self,arr,num,diff):
    self.arr=arr
    self.num=num
    self.diff=diff
  def find(self):
    count=0
    for i in range(len(self.arr)):
      if(abs(self.arr[i]-num)<=diff):
        count+=1
    return count
arr=list(map(int,input().split()))
num=int(input())
diff=int(input())
print(comb_count(arr,num,diff).find())

12 3 14 56 77 13
13 
2
3


In [ ]:
class comb_count:
  @classmethod
  def find(cls,arr,num,diff):
    count=0
    for i in range(len(arr)):
      if(abs(arr[i]-num)<=diff):
        count+=1
    return count
arr=list(map(int,input().split()))
num=int(input())
diff=int(input())
print(comb_count().find(arr,num,diff))


12 3 14 56 77 13 
13
2
3




```
n=4
m=20
o/p=90
sum of nums divisible by n upto m limit=(4 8 12 16 20)=>>60
non divisiors=(1 2 3 ......19)=>>150

diff=150-60 ==>90
```



In [ ]:
def find(n,m):
  divisors=[]
  non_divisors=[]
  for i in range(1,m+1):
    if i%n==0:
      divisors.append(i)
    else:
      non_divisors.append(i)
  return divisors,non_divisors
n=int(input())
m=int(input())
l1,l2=find(n,m)
a=sum(l1)
b=sum(l2)
diff=a-b
print(abs(diff))



4
20
90


In [ ]:
class find_diff:
    def __init__(self, n, m):
        self.n = n
        self.m = m

    def find(self):
        divisors = []
        non_divisors = []
        for i in range(1, self.m + 1):
            if i % self.n == 0:
                divisors.append(i)
            else:
                non_divisors.append(i)
        return divisors, non_divisors
n = int(input())
m = int(input())
a,b= find_diff(n, m).find()
a=sum(a)
b=sum(b)
diff=a-b
print(abs(diff))



4
20


TypeError: 'int' object is not callable

In [ ]:
class find_diff:
  @classmethod
  def find(cls,n,m):
    divisors=[]
    non_divisors=[]
    for i in range(1,m+1):
      if i%n==0:
        divisors.append(i)
      else:
        non_divisors.append(i)
    return sum(divisors)-sum(non_divisors)
n=int(input())
m=int(input())
diff=find_diff().find(n,m)
print(abs(diff))


4
20


TypeError: 'int' object is not callable



```
i/p:
arr=[3,2,1,7,5,4]
o/p:7

even_pos_arr=[3,1,5] =>sort=[1,3,5]
 2nd largest=3
odd pos arr=[2,7,4] =>sort=[2,4,7]
 2nd smallest element=4
 sum=3+4=7 =>output

```



In [ ]:
def find_sum(arr):
  even_list=[]
  odd_list=[]
  for i in range(len(arr)):
    if i%2==0:
      even_list.append(arr[i])
    else:
      odd_list.append(arr[i])
  return sorted(even_list),sorted(odd_list)
arr=list(map(int,input().split()))
list1,list2=find_sum(arr)
print(list1[len(list1)-2]+list2[1])





3 2 1 7 5 4
7


In [ ]:
class arr_sum:
  def __init__(self,arr):
    self.arr=arr
  def find_sum(self):
    even_list=[]
    odd_list=[]
    for i in range(len(self.arr)):
      if i%2==0:
        even_list.append(self.arr[i])
      else:
        odd_list.append(self.arr[i])
    return sorted(even_list),sorted(odd_list)
arr=list(map(int,input().split()))
l1,l2=arr_sum(arr).find_sum()
print(l1[len(l1)-2]+l2[1])

3 2 1 7 5 4
7


In [ ]:
class arr_sum:
  @classmethod
  def find_sum(cls,arr):
    even_list=[]
    odd_list=[]
    for i in range(len(arr)):
      if i%2==0:
        even_list.append(arr[i])
      else:
        odd_list.append(arr[i])
    return sorted(even_list),sorted(odd_list)
arr=list(map(int,input().split()))
l1,l2=arr_sum().find_sum(arr)
print(l1[len(l1)-2]+l2[1])


3 2 1 7 4  5
8




```
sum=9
size of arr=7
arr=5 2 4 3 9 7 1
o/p=2
pair of least 2 element(2,1)=2+1=3=> 3<9
so 2*1=2 is ouput
```



In [ ]:
def find(sum,size,arr):
  a=sorted(set(arr))
  product=1
  if(a[0]+a[1])<sum:
    product=a[0]*a[1]
  return product
sum=int(input())
size=int(input())
arr=list(map(int,input().split()))
res=find(sum,size,arr)
print(res)


4
6
12 9 -7 -7 3 8
-21


In [ ]:
class least_pair_product:
  def __init__(self,sum,size,arr):
    self.sum=sum
    self.size=size
    self.arr=arr
  def find(self):
    a=sorted(set(self.arr))
    product=1
    if(a[0]+a[1]<self.sum):
      product=a[0]*a[1]
    return product
sum=int(input())
size=int(input())
arr=list(map(int,input().split()))
res=least_pair_product(sum,size,arr).find()
print(res)



9
7
5 4 2 3 1 7 9
2


In [ ]:
class least_pair_product:
  @classmethod
  def find(cls,sum,size,arr):
    a=sorted(set(arr))
    product=1
    if(a[0]+a[1])<sum:
      product=a[0]*a[1]
    return product
sum=int(input())
size=int(input())
arr=list(map(int,input().split()))
res=least_pair_product().find(sum,size,arr)
print(res)


9
7
5 4 2 3 1 7 9
2




```
INPUT1=A B C E
INPUT2=B C D
O/P=8

```



In [ ]:
def find(set1,set2):
  list1=[]
  uncommon_set=set1-set2
  tot_sum=0
  for i in uncommon_set:
    list1.append(ord(i))
    tot_sum=sum(list1)
  return tot_sum
def find_digit_sum(res):
  while res>9:
    digit_sum=0
    while res>0:
      digit=res%10
      digit_sum=digit_sum+digit
      res=res//10
    res=digit_sum
  return res
set1=set(input().split())
set2=set(input().split())
res=find(set1,set2)
res1=find_digit_sum(res)
print(res1)



X Y Z
A B
6


In [ ]:
class single_digit_sum:
  def __init__(self,set1,set2):
    self.set1=set1
    self.set2=set2
  def find(self):
    list1=[]
    uncommon_set=self.set1-self.set2
    tot_sum=0
    for i in uncommon_set:
      list1.append(ord(i))
      tot_sum=sum(list1)
    return tot_sum
  def find_digit_sum(self,res):
    while res>9:
      digit_sum=0
      while res>0:
        digit=res%10
        digit_sum=digit_sum+digit
        res=res//10
      res=digit_sum
    return res
set1=set(input().split())
set2=set(input().split())
obj=single_digit_sum(set1,set2)
res=obj.find()
res1=obj.find_digit_sum(res)
print(res1)


X Y Z
A B
6


In [ ]:
class single_digit_sum:
  @classmethod
  def find(cls,set1,set2):
    tot_sum=0
    uncommon_set=set1-set2
    list1=[]
    for i in uncommon_set:
      list1.append(ord(i))
      tot_sum=sum(list1)
    return tot_sum
  @classmethod
  def find_digit_sum(cls,res):
    while res>9:
      digit_sum=0
      while res>0:
        digit=res%10
        digit_sum=digit_sum+digit
        res=res//10
      res=digit_sum
    return res
  @classmethod
  def display(cls,set1,set2):
    res=cls.find(set1,set2)
    res1=cls.find_digit_sum(res)
    print(res1)
set1=set(input().split())
set2=set(input().split())
single_digit_sum().display(set1,set2)





X Y Z
A B
6


In [ ]:
def find(set1,set2):
  uncommon_set=set1-set2
  tot_sum=sum(ord(i) for i in uncommon_set)
  return tot_sum
def find_digit_sum(digit_sum):
  while digit_sum>9:
    digit_sum=sum(int(i) for i in str(digit_sum))
  return digit_sum
set1=set(input().split())
set2=set(input().split())
set_sum=find(set1,set2)
digit_sum=find_digit_sum(set_sum)
print(digit_sum)



X Y Z
A B
6


In [ ]:
class single_digit_sum:
  def __init__(self,set1,set2):
    self.set1=set1
    self.set2=set2
  def find(self):
    uncommon_set=self.set1-self.set2
    tot_sum=sum(ord(i) for i in uncommon_set)
    return tot_sum
  def find_digit_sum(self,digit_sum):
    while digit_sum>9:
      digit_sum=sum(int(i) for i in str(digit_sum))
    return digit_sum
set1=set(input().split())
set2=set(input().split())
set_sum=single_digit_sum(set1,set2).find()
digit_sum=single_digit_sum(set1,set2).find_digit_sum(set_sum)
print(digit_sum)

X Y Z
A B
6


In [ ]:
class single_digit_sum:
  @classmethod
  def find(cls,set1,set2):
    uncommon_set=set1-set2
    tot_sum=sum(ord(i) for i in uncommon_set)
    return tot_sum
  @classmethod
  def find_digit_sum(cls,digit_sum):
    while digit_sum>9:
      digit_sum=sum(int(i) for i in str(digit_sum))
    return digit_sum
  @classmethod
  def display(cls,set1,set2):
    set_sum=cls.find(set1,set2)
    digit_sum=cls.find_digit_sum(set_sum)
    print(digit_sum)
set1=set(input().split())
set2=set(input().split())
single_digit_sum().display(set1,set2)

X Y Z
A B
6


**FASCINATING NUMBER:**
A number is called fascinating if after multiplying it by 2 and 3 and concatenating the results with the original number, all digits from 1 to 9 appear exactly once.

Example:
Consider 192

192 × 1 = 192

192 × 2 = 384

192 × 3 = 576

Concatenated: 192384576

It contains all digits from 1 to 9 exactly once → ✅ Fascinating!

In [ ]:
def facinating(n):
  if n<100:
    return False
  concat=str(n)+str(n*2)+str(n*3)
  digits=sorted(concat)
  num_list=['1','2','3','4','5','6','7','8','9']
  return digits==num_list
n=int(input())
if facinating(n):
  print("fascinating num!")
else:
  print("not a fascinating num!")


273
fascinating num!


In [ ]:
class fascinating:
  def __init__(self,n):
    self.n=n
  def find(self):
    if self.n<100:
      return False
    concat=str(self.n)+str(self.n*2)+str(self.n*3)
    num_list=['1','2','3','4','5','6','7','8','9']
    return sorted(concat)==num_list
n=int(input())
obj=fascinating(n)
if obj.find():
  print("fascinating num!")
else:
  print("not a fascinating num!")




219
fascinating num!


In [ ]:
class fascinating:
  @classmethod
  def find(cls,n):
    if n<100:
      return False
    concat=str(n)+str(n*2)+str(n*3)
    num_list=['1','2','3','4','5','6','7','8','9']
    return sorted(concat)==num_list
  @classmethod
  def display(cls,n):
    if cls.find(n):
      print("fascinating num!")
    else:
      print("not a fascinating num!")
n=int(input())
fascinating().display(n)



273
fascinating num!


In [ ]:
def fascinating(n):
  if n<100:
    return False
  concat=str(n)+str(n*2)+str(n*3)
  return sorted(concat)==list("123456789")
start=int(input())
end=int(input())
fas_nums=[i for i in range(start,end+1) if fascinating(i)]
if fas_nums:
  print(*fas_nums)
else:
  print("no fascinating num in range!")

100
1000
192 219 273 327


In [ ]:
class fascinating:
  def __init__(self,start,end):
    self.start=start
    self.end=end
  def check(self,n):
    if n<100:
      return False
    concat=str(n)+str(n*2)+str(n*3)
    return sorted(concat)==list("123456789")
  def find_range(self,start,end):
    return [i for i in range(self.start,self.end+1) if self.check(i)]
  def display(self):
    fas_num=self.find_range(self.start,self.end)
    if fas_num:
      print(*fas_num)
    else:
      print("no fascinating numbers in the range!")
start=int(input())
end=int(input())
obj=fascinating(start,end)
obj.display()


100
1000
192 219 273 327


In [ ]:
class fascinating_num:
  @classmethod
  def check(cls,n):
    if n<100:
      return False
    concat=str(n)+str(n*2)+str(n*3)
    return sorted(concat)==list("123456789")
  @classmethod
  def find_range(cls,start,end):
    return [i for i in range(start,end+1) if cls.check(i)]
  @classmethod
  def display(cls,start,end):
    fas_num=cls.find_range(start,end)
    if fas_num:
      print(*fas_num)
    else:
      print("no fascinating numbers in range")
start=int(input())
end=int(input())
fascinating_num().display(start,end)



100
1000
192 219 273 327


**PERFECT NUMBER:**
A Perfect Number is a positive integer that is equal to the sum of its proper positive divisors, excluding itself.

For example, 6 is a perfect number
1 + 2 + 3 = 6

In [ ]:
def perfect_num(n):
  if n<=1:
    return False
  sum_divisors=0
  for i in range(1,n):
    if n%i==0:
      sum_divisors+=i
  return sum_divisors==n
n=int(input())
if perfect_num(n):
  print("perfect num!")
else:
  print("not a perfect num!")


212
not a perfect num!


In [ ]:
class perfect_num:
  def __init__(self,n):
    self.n=n
  def find(self):
    if self.n<=1:
      return False
    sum_divisors=0
    for i in range(1,self.n):
      if self.n%i==0:
        sum_divisors+=i
    return sum_divisors==n
  def display(self):
    if self.find():
      print("perfect num!")
    else:
      print("not a perfect num!")
n=int(input())
perfect_num(n).display()



6
perfect num!


In [ ]:
class perfect_num:
  @classmethod
  def find(cls,n):
    if n<=1:
      return False
    sum_div=0
    for i in range(1,n):
      if n%i==0:
        sum_div+=i
    return sum_div==n
  @classmethod
  def display(cls,n):
    if cls.find(n):
      print("perfect num!")
    else:
      print("not perfect num!")
n=int(input())
perfect_num().display(n)

28
perfect num!


perfect num in range

In [ ]:
def perfect_range(n):
  if n<=1:
    return False
  return sum(i for i in range(1,n) if n%i==0)==n
def find_range(start,end):
  return [i for i in range(start,end) if perfect_range(i)]
def display(start,end):
  perfect_num=find_range(start,end)
  if perfect_num:
    print(*perfect_num)
  else:
    print("no perfect num in range!!")
start=int(input())
end=int(input())
display(start,end)


1
30
6 28


In [ ]:
class perfect_numbers:
  def __init__(self,start,end):
    self.start=start
    self.end=end
  def check_perfect(self,n):
    if n<=1:
      return False
    return sum(i for i in range(1,n) if n%i==0)==n
  def find_range(self,start,end):
    return [i for i in range(self.start,self.end) if self.check_perfect(i)]
  def display(self):
    nums=self.find_range(self.start,self.end)
    if nums:
      print(*nums)
    else:
      print("no perfect num in the range!")
start=int(input())
end=int(input())
perfect_numbers(start,end).display()



1
100
6 28


In [ ]:
class perfect_nums:
  @classmethod
  def check_perfect(cls,n):
    if n<=1:
      return False
    return sum(i for i in range(1,n) if n%i==0)==n
  @classmethod
  def find_range(cls,start,end):
    return [i for i in range(start,end) if cls.check_perfect(i)]
  @classmethod
  def display(cls,start,end):
    nums=cls.find_range(start,end)
    if nums:
      print(*nums)
    else:
      print("no perfect num in the range!")
start=int(input())
end=int(input())
perfect_nums.display(start,end)


1
4
no perfect num in the range!


palindrome -string

In [ ]:
class palindrome_str:
  @classmethod
  def check(cls,s):
    s=s.lower()
    left=0
    right=len(s)-1
    while left<right:
      if s[left]!=s[right]:
        return False
      left+=1
      right-=1
    return True
  @classmethod
  def display(cls,s):
    if cls.check(s):
      print("2")
    else:
      print("1")

s=input()
palindrome_str.display(s)


house
1


In [ ]:
def palindrome_check(s):
  s=s.lower()
  left=0
  right=len(s)-1
  while left<right:
    if s[left]!=s[right]:
      return False
    left+=1
    right-=1
  return True
s=input()
if palindrome_check(s):
  print("palindrome!")
else:
  print("not a plaindrome!!")


pop
palindrome!


In [ ]:
class palindrome:
  def __init__(self,s):
    self.s=s
  def check(self):
    self.s=self.s.lower()
    left=0
    right=len(self.s)-1
    while left<right:
      if self.s[left]!=self.s[right]:
        return False
      left+=1
      right-=1
    return True
  def display(self):
    if self.check() :
      print("palindrome!")
    else:
      print("not palindrome!")
s=input()
palindrome(s).display()

appa
palindrome!


In [ ]:
class palindrome:
  @classmethod
  def check(cls,s):
    s=s.lower()
    left=0
    right=len(s)-1
    while left<right:
      if s[left]!=s[right]:
        return False
      left+=1
      right-=1
    return True
  @classmethod
  def display(cls,s):
    if cls.check(s):
      print("palindrome!!")
    else:
      print("not a palindrome!")
s=input()
palindrome.display(s)


amma
palindrome!!


if palindrome return 2 else return 1 (for both str and num)

In [ ]:
class palindrome_str:
    @classmethod
    def check_and_display(cls, s):
        s = s.lower()
        left = 0
        right = len(s) - 1
        while left < right:
            if s[left] != s[right]:
                return 1
            left += 1
            right -= 1
        return 2
s = input()
palindrome_str.check_and_display(s)


1221


2

In [ ]:
class palindrome_str:
    @classmethod
    def check(cls, input1):
        input1 = input1.lower()
        s2 = input1[::-1]
        for i in range(len(input1)):
            if input1[i] != s2[i]:
                return 1
        return 2
s = input()
print(palindrome_str.check(s))


pop
2


In [ ]:
def palindrome(s):
  s=s.lower()
  s1=s[::-1]
  for i in range(len(s)):
    if s[i]!=s1[i]:
      return False
  return True
s=input()
if palindrome(s):
  print("palindrome!")
else:
  print("not a palindrome!")


1221
palindrome!


PALINDROME -NUMBER

In [ ]:
def palindrome_num(n):
  temp=n
  reverse_n=0
  while n>0:
    digit=n%10
    reverse_n=reverse_n*10+digit
    n//=10
  return reverse_n==temp
n=int(input())
if palindrome_num(n):
  print("palindrome!")
else:
  print("not a palindrome!")

9009
palindrome!


In [ ]:
class palindrome:
  def __init__(self,n):
    self.n=n
  def check(self):
    temp=self.n
    reverse_n=0
    while self.n>0:
      digit=self.n%10
      reverse_n=reverse_n*10+digit
      self.n//=10
    return temp==reverse_n
  def display(self):
    if self.check():
      print("palindrome!!")
    else:
      print("not a palindrome!")
n=int(input())
palindrome(n).display()

1001
palindrome!!


In [ ]:
class palindrome:
  @classmethod
  def check(cls,n):
    temp=n
    reverse_n=0
    while n>0:
      digit=n%10
      reverse_n=reverse_n*10+digit
      n//=10
    return temp==reverse_n
  @classmethod
  def display(cls,n):
    if cls.check(n):
      print("palindrome!!")
    else:
      print("not a plaindrome!!")
n=int(input())
palindrome.display(n)

3003
palindrome!!


**ARMSTRONG NUMBER:**
An Armstrong number (also called a narcissistic number) is a number that is equal to the sum of its digits each raised to the power of the number of digits.

Examples:

153 → 1³ + 5³ + 3³ = 153 → ✅ Armstrong

9474 → 9⁴ + 4⁴ + 7⁴ + 4⁴ = 9474 → ✅ Armstrong


In [ ]:
def armstrong(n):
  temp=n
  sum_of_pow=0
  digits=len(str(n))
  while n>0:
    digit=n%10
    sum_of_pow+=digit**digits
    n//=10
  return sum_of_pow==temp
n=int(input())
if armstrong(n):
  print("armstrong num !")
else:
  print("not armstrong!")


153
armstrong num !


In [ ]:
class armstrong:
  def __init__(self,n):
    self.n=n
  def check(self):
    temp=self.n
    sum_of_pow=0
    digits=len(str(self.n))
    while self.n>0:
      digit=self.n%10
      sum_of_pow+=digit**digits
      self.n//=10
    return temp==sum_of_pow
  def display(self):
    if self.check():
      print("armstrong!")
    else:
      print("not armstrong!")
n=int(input())
armstrong(n).display()


153
armstrong!


In [ ]:
class armstrong:
  @classmethod
  def check(cls,n):
    temp=n
    sum_of_pow=0
    digits=len(str(n))
    while n>0:
      digit=n%10
      sum_of_pow+=digit**digits
      n//=10
    return temp==sum_of_pow
  @classmethod
  def display(cls,n):
    if cls.check(n):
      print("armstrong!")
    else:
      print("not armstrong!")
n=int(input())
armstrong.display(n)


234
not armstrong!


ARMSTRONG NUM IN RANGE

In [ ]:
def armstrong_check(n):
  temp=n
  sum_of_pow=0
  digits=len(str(n))
  while n>0:
    digit=n%10
    sum_of_pow+=digit**digits
    n//=10
  return temp==sum_of_pow
start=int(input())
end=int(input())
armstrong_nums=[i  for i in range(start,end) if armstrong_check(i)]
if armstrong_nums:
  print(*armstrong_nums)
else:
  print("no armstrong num in range!")

100
1000
153 370 371 407


In [ ]:
class armstrong:
  def __init__(self,start,end):
    self.start=start
    self.end=end
  def check(self,n):
    temp=n
    sum_of_pow=0
    digits=len(str(n))
    while n>0:
      digit=n%10
      sum_of_pow+=digit**digits
      n//=10
    return temp==sum_of_pow
  def find_range(self,start,end):
    return [i for i in range(self.start,self.end) if self.check(i)]
  def display(self):
    nums=self.find_range(self.start,self.end)
    if nums:
      print(*nums)
    else:
      print("no armstrong in the range!")
start=int(input())
end=int(input())
armstrong(start,end).display()

100
1000
153 370 371 407


In [ ]:
class armstrong:
  @classmethod
  def check(cls,n):
    temp=n
    sum_of_pow=0
    digits=len(str(n))
    while n>0:
      digit=n%10
      sum_of_pow+=digit**digits
      n//=10
    return temp==sum_of_pow
  @classmethod
  def find_range(cls,start,end):
    return [i for i in range(start,end) if cls.check(i)]
  @classmethod
  def display(cls,start,end):
    nums=cls.find_range(start,end)
    if nums:
      print(*nums)
    else:
      print("no armstrong in the range!")
start=int(input())
end=int(input())
armstrong.display(start,end)


100
1000
153 370 371 407


**FIBONACCI SERIES:**
The Fibonacci series is a sequence of numbers in which each number is the sum of the two preceding ones, starting from 0 and 1.

F(0) = 0

F(1) = 1

F(n) = F(n−1) + F(n−2) for n ≥ 2

In [ ]:
def fibonacci(n):
  n1=0
  n2=1
  for i in range(1,n):
    print(n1)
    n3=n1+n2
    n1=n2
    n2=n3
n=int(input())
fibonacci(n)


10
0
1
1
2
3
5
8
13
21


In [ ]:
class fibo:
  def __init__(self,n):
    self.n=n
  def display(self):
    n1=0
    n2=1
    for i in range(self.n):
      print(n1)
      n3=n1+n2
      n1=n2
      n2=n3
n=int(input())
fibo(n).display()

10
0
1
1
2
3
5
8
13
21
34


In [ ]:
class fibo:
  @classmethod
  def display(cls,n):
    n1=0
    n2=1
    for i in range(n):
      print(n1)
      n3=n1+n2
      n1=n2
      n2=n3
n=int(input())
fibo.display(n)



5
0
1
1
2
3


In [ ]:
def fibonacci(n,m):
  n1=0
  n2=1
  while n1<=m:
    if n1>=n:
      print(n1)
    n3=n1+n2
    n1=n2
    n2=n3
n=int(input())
m=int(input())
fibonacci(n,m)

5
15
5
8
13


In [ ]:
class fibo:
  def __init__(self,n,m):
    self.n=n
    self.m=m
  def display(self):
    n1=0
    n2=1
    while n1<=self.m:
      if n1>=self.n:
        print(n1)
      n3=n1+n2
      n1=n2
      n2=n3
n=int(input())
m=int(input())
fibo(n,m).display()




5
15
5
8
13


In [ ]:
class fibo:
  @classmethod
  def display(cls,n,m):
    n1=0
    n2=1
    while n1<=m:
      if n1>=n:
        print(n1)
      n3=n1+n2
      n1=n2
      n2=n3
n=int(input())
m=int(input())
fibo.display(n,m)



5
20
5
8
13


**COMPOSITE NUMBERS:**
a number that has more than two factors, i.e., not prime and greater than 1


In [ ]:
def composite(n):
  if n<=1:
    return False
  for i in range(2,n):
    if n%i==0:
      return True
  return False
def display(n):
  if composite(n):
    print("it is composite number!!")
  else:
    print("not a composite number!!")
n=int(input())
display(n)

2
not a composite number!!


In [ ]:
class compo:
  def __init__(self,n):
    self.n=n
  def check(self):
    if n<=1:
      return False
    for i in range(2,n):
      if n%i==0:
        return True
    return False
  def display(self):
    if self.check():
      print("composite num!")
    else:
      print("not a composite num!")
n=int(input())
compo(n).display()


15
composite num!


In [ ]:
class compo:
  @classmethod
  def check(cls,n):
    if n<=1:
      return False
    for i in range(2,n):
      if n%i==0:
        return True
    return False
  @classmethod
  def display(cls,n):
    if cls.check(n):
      print("composite num!")
    else:
      print("not a composite num!")
n=int(input())
compo.display(n)

11
not a composite num!


In [ ]:
def composite(n):
  if n<=1:
    return False
  for i in range(2,n):
    if n%i==0:
      return True
  return False
def display(start,end):
  return [i for i in range(start,end) if composite(i)]
start=int(input())
end=int(input())
comp_list=display(start,end)
if comp_list:
  print(*comp_list)
else:
  print("no composite numbers in range")

1
20
4 6 8 9 10 12 14 15 16 18


In [ ]:
class comp:
  def __init__(self,start,end):
    self.start=start
    self.end=end
  def check(self,n):
    if n<=1:
      return False
    for i in range(2,n):
      if n%i==0:
        return True
    return False
  def display(self):
    return [i for i in range(self.start,self.end) if self.check(i)]
start=int(input())
end=int(input())
comp_list=comp(start,end).display()
if comp_list:
  print(*comp_list)
else:
  print("no composite num in range!")



2
20
4 6 8 9 10 12 14 15 16 18


In [ ]:
class compo:
  @classmethod
  def check(cls,n):
    if n<=1:
      return False
    for i in range(2,n):
      if n%i==0:
        return True
    return False
  @classmethod
  def display(cls,start,end):
    return [i for i in range(start,end) if cls.check(i)]
start=int(input())
end=int(input())
comp_list=compo.display(start,end)
if comp_list:
  print(*comp_list)
else:
  print("no composite num in range!")


2
10
4 6 8 9


**PRIME AND COMPOSITE NUMBER CHECK:**

In [ ]:
def check(n):
  if n<=1:
    return False
  for i in range(2,n):
    if n%i==0:
      return True
  return False
n=int(input())
if check(n):
  print("composite number!")
else:
  print("prime number!")


21
composite number!


In [ ]:
class comp:
  def __init__(self,n):
    self.n=n
  def check(self):
    if self.n<=1:
      return False
    for i in range(2,self.n):
      if self.n%i==0:
        return True
    return False
  def display(self):
    if self.check():
      print("it is composite num!")
    else:
      print("it is not a composite num!")
n=int(input())
comp(n).display()

54
it is composite num!


In [ ]:
class compo:
  @classmethod
  def check(cls,n):
    if n<=1:
      return False
    for i in range(2,n):
      if n%i==0:
        return True
    return False
  @classmethod
  def display(cls,n):
    if cls.check(n):
      print("it is composite num!")
    else:
      print("it is not a composite num!")
n=int(input())
compo.display(n)

3
it is not a composite num!


take int array as user input.print prime and composite numbers in separte array

In [ ]:
def comp_prime(array):
  prime=[]
  comp=[]
  for i in array:
    if i<=1:
      continue
    is_prime=True
    for j in range(2,i):
      if i%j==0:
        is_prime=False
        break
    if is_prime:
      prime.append(i)
    else:
      comp.append(i)
  return prime,comp

size=int(input())
array=[int(input()) for i in range(size)]
prime_arr,comp_arr=comp_prime(array)
print("prime numbers:",*prime_arr)
print("composite numbers:",*comp_arr)


5
12
65
1
3
7
prime numbers: 3 7
composite numbers: 12 65


In [ ]:
class comp_prime:
  def __init__(self,arr):
    self.arr=arr
  def check(self):
    prime=[]
    comp=[]
    for i in self.arr:
      if i<=1:
        continue
      is_prime=True
      for j in range(2,i):
        if i%j==0:
          is_prime=False
          break
      if is_prime:
        prime.append(i)
      else:
       comp.append(i)
    return prime,comp
  def display(self):
    prime_arr,comp_arr=self.check()
    print("prime num:",*prime_arr)
    print("composite num:",*comp_arr)
size=int(input())
arr=[int(input()) for i in range(size)]
comp_prime(arr).display()

5
32
7
87
12
5
prime num: 7 5
composite num: 32 87 12


In [ ]:
class compo_prime:
  @classmethod
  def check(cls,arr):
    prime=[]
    comp=[]
    for i in arr:
      if i<=1:
        continue
      is_prime=True
      for j in range(2,i):
        if i%j==0:
          is_prime=False
          break
      if is_prime:
        prime.append(i)
      else:
        comp.append(i)
    return prime,comp
  @classmethod
  def display(cls,arr):
    prime_arr,comp_arr=cls.check(arr)
    print("prime numbers:",*prime_arr)
    print("composite numbers:",*comp_arr)
size=int(input())
arr=[int(input()) for i in range(size)]
compo_prime.display(arr)

4
21
32
47
9
prime numbers: 47
composite numbers: 21 32 9


taking range from user input and printing prime and composite numbers in that range

In [ ]:
def comp_prime(arr):
  prime=[]
  comp=[]
  for i in arr:
    if i<=1:
      continue
    is_prime=True
    for j in range(2,i):
      if i%j==0:
        is_prime=False
        break
    if is_prime:
      prime.append(i)
    else:
      comp.append(i)
  return prime,comp
def display(arr):
  prime_arr,comp_arr=comp_prime(arr)
  print("prime num:",*prime_arr)
  print("composite num:",*comp_arr)
start=int(input())
end=int(input())
arr=[i for i in range(start,end)]
display(arr)


4
20
prime num: 5 7 11 13 17 19
composite num: 4 6 8 9 10 12 14 15 16 18


In [ ]:
class comp_prime:
  def __init__(self,arr):
    self.arr=arr
  def check(self):
    prime=[]
    comp=[]
    for i in self.arr:
      if i<=1:
        continue
      is_prime=True
      for j in range(2,i):
        if i%j==0:
          is_prime=False
          break
      if is_prime:
        prime.append(i)
      else:
        comp.append(i)
    return prime,comp
  def display(self):
    prime_arr,comp_arr=self.check()
    print("prime num:",*prime_arr)
    print("comp num:",*comp_arr)
start=int(input())
end=int(input())
arr=[i for i in range(start,end)]
comp_prime(arr).display()


10
25
prime num: 11 13 17 19 23
comp num: 10 12 14 15 16 18 20 21 22 24


In [ ]:
class prime_comp:
  @classmethod
  def check(cls,arr):
    prime=[]
    comp=[]
    for i in arr:
      if i<=1:
        continue
      is_prime=True
      for j in range(2,i):
        if i%j==0:
          is_prime=False
          break
      if is_prime:
        prime.append(i)
      else:
        comp.append(i)
    return prime,comp
  @classmethod
  def display(cls,arr):
    prime_arr,comp_arr=cls.check(arr)
    print("prime num:",*prime_arr)
    print("composite numbers:",*comp_arr)
start=int(input())
end=int(input())
arr=[i for i in range(start,end)]
prime_comp.display(arr)



25
100
prime num: 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97
composite numbers: 25 26 27 28 30 32 33 34 35 36 38 39 40 42 44 45 46 48 49 50 51 52 54 55 56 57 58 60 62 63 64 65 66 68 69 70 72 74 75 76 77 78 80 81 82 84 85 86 87 88 90 91 92 93 94 95 96 98 99


SUM OF DIGITS

In [ ]:
def sum_of_digits(n):
  sum=0
  while n>0:
    digit=n%10
    sum+=digit
    n//=10
  return sum
n=int(input())
print(sum_of_digits(n))

1234
10


In [ ]:
class sum_of_digit:
  def __init__(self,n):
    self.n=n
  def find(self):
    sum=0
    while self.n>0:
      digit=self.n%10
      sum+=digit
      self.n//=10
    return sum
n=int(input())
print(sum_of_digit(n).find())

145
10


In [ ]:
class sum_of_digit:
  @classmethod
  def find(cls,n):
    sum=0
    while n>0:
      digit=n%10
      sum+=digit
      n//=10
    return sum
n=int(input())
print(sum_of_digit.find(n))

35
8


LCM OF TWO INTEGERS

In [ ]:
def lcm(a,b):
  max_n=max(a,b)
  while True:
    if max_n%a==0 and max_n%b==0:
      return max_n
    max_n+=1
a=int(input())
b=int(input())
print("LCM:",lcm(a,b))

4
6
LCM: 12


In [ ]:
class lcm:
  def __init__(self,a,b):
    self.a=a
    self.b=b
  def find(self):
    max_n=max(self.a,self.b)
    while True:
      if max_n%self.a==0 and max_n%self.b==0:
        return max_n
      max_n+=1
a=int(input())
b=int(input())
print("lcm:",lcm(a,b).find())

7
5
lcm: 35


In [ ]:
class lcm:
  @classmethod
  def find(cls,a,b):
    max_n=max(a,b)
    while True:
      if max_n%a==0 and max_n%b==0:
        return max_n
      max_n+=1
a=int(input())
b=int(input())
print("lcm:",lcm.find(a,b))

7
5
lcm: 35


GCD(GREATEST COMMON DIVISIOR )/ HCF

GCD of two numbers is the largest number that divides both numbers exactly (without a remainder).

In [ ]:
def gcd(a,b):
  while b!=0:
    temp=a
    a=b
    b=temp%b
  return a
a=int(input())
b=int(input())
print("GCD:",gcd(a,b))

18
24
GCD: 6


In [ ]:
class gcd:
  def __init__(self,a,b):
    self.a=a
    self.b=b
  def find(self):
    while self.b!=0:
      temp=self.a
      self.a=self.b
      self.b=temp%self.b
    return self.a
a=int(input())
b=int(input())
print("gcd:",gcd(a,b).find())



18
24
gcd: 6


In [ ]:
class gcd:
  @classmethod
  def find(cls,a,b):
    while b!=0:
      temp=a
      a=b
      b=temp%b
    return a
a=int(input())
b=int(input())
print("gcd:",gcd.find(a,b))

12
18
gcd: 6


In [ ]:
def gcd(a,b):
  while b:
    a,b=b,a%b
  return a
a=int(input())
b=int(input())
print("gcd:",gcd(a,b))


3
5
gcd: 1


REVERSE NUMBER

In [ ]:
def reverse(n):
  reversed=0
  while n>0:
    digit=n%10
    reversed=reversed*10+digit
    n//=10
  return reversed
n=int(input())
print("reversed number:",reverse(n))

2154
reversed number: 4512


In [ ]:
def reverse(n):
  reversed=0
  while n>0:
    reversed=reversed*10+(n%10)
    n//=10
  return reversed
n=int(input())
print("reversed number:",reverse(n))

31654
reversed number: 45613


In [ ]:
class reverse:
  def __init__(self,n):
    self.n=n
  def find(self):
    reversed=0
    while self.n>0:
      reversed=reversed*10+(self.n % 10)
      self.n//=10
    return reversed
n=int(input())
print("reversed number:",reverse(n).find())


348
reversed number: 843


In [ ]:
class reverse:
  @classmethod
  def find(cls,n):
    reversed=0
    while n>0:
      reversed=reversed*10 +(n%10)
      n//=10
    return reversed
  @classmethod
  def display(cls,n):
    print("reverse num:",reverse.find(n))
n=int(input())
reverse.display(n)

345
reverse num: 543


REVERSING NUM AND STRING

In [ ]:
def check(s1):
  return s1.lower()[::-1]
s=input()
print("reversing str:",check(s))

knjbhv524
reversing str: 425vhbjnk


In [ ]:
class reverse:
  def __init__(self,s):
    self.s=s
  def find(self):
    return self.s.lower()[::-1]
s=input()
print("reversing str:",reverse(s).find())

354
reversing str: 453


In [ ]:
class reverse:
  @classmethod
  def find(cls,s):
    return s.lower()[::-1]
s=input()
print("reversing string:",reverse.find(s))

54578
reversing string: 87545


REVERSING STRING WITHOUT USING BULIT IN FUNCTION:


In [ ]:
def reverse(s):
  res=""
  for i in s:
    res=i+res
  return res
s=input()
print(reverse(s))

kj65
56jk


In [ ]:
def reverse(s):
  return ''.join(i for i in s)[::-1]
s=input()
print("reverse:",reverse(s))

ninja
reverse: ajnin


ODD AND EVEN NUMBER

In [ ]:
n=int(input("enter number:"))
if n%2==0:
  print("even num!")
else:
  print("odd num!")


enter number:34
even num!


In [ ]:
def odd_even(n):
  return "even num!" if n%2==0 else "odd num!"
n=int(input("enter num:"))
print(odd_even(n))

enter num:321
odd num!


In [ ]:
class even_odd:
  def __init__(self,n):
    self.n=n
  def check(self):
    return "even num!" if self.n%2==0 else "odd num!"
n=int(input("enter num:"))
print(even_odd(n).check())


enter num:65
odd num!


In [ ]:
class evn_odd:
  @classmethod
  def check(cls,n):
    return "even num" if n%2==0 else "odd_num"
n=int(input())
print(evn_odd.check(n))

98
even num


In [ ]:
#lambda function
even_odd_check=lambda x:"even" if x%2==0 else "odd"
n=int(input())
print(even_odd_check(n))

18
even


In [ ]:
#match case
def even_odd(n):
  match n%2:
    case 0:
      return "even"
    case 1:
      return "odd"
n=int(input())
print(even_odd(n))

23
odd


In [ ]:
#list of even and odd numbers
size=int(input())
arr=[int(input()) for i in range(size)]
def even_odd(arr):
  even=[]
  odd=[]
  for i in arr:
    if i%2==0:
      even.append(i)
    else:
      odd.append(i)
  return even,odd
even_lst,odd_lst=even_odd(arr)
print("even numbers:",even_lst)
print("odd numbers:",odd_lst)

5
2
34
12
61
63
even numbers: [2, 34, 12]
odd numbers: [61, 63]


In [ ]:
#start and end
start=int(input())
end=int(input())
arr=[i for i in range(start,end)]
def even_odd(arr):
  even=[]
  odd=[]
  for i in arr:
    if i%2==0:
      even.append(i)
    else:
      odd.append(i)
  return even,odd
even_lst,odd_lst=even_odd(arr)
print("even numbers:",even_lst)
print("odd numbers:",odd_lst)




PUNCTATION REMOVING

In [ ]:
import string
punctuation_marks=string.punctuation
print(punctuation_marks)
def remove_punct(text):
  return text.translate(str.maketrans('','',punctuation_marks))
text="helloo!!,.This is 3sha ####"
print(remove_punct(text))

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~
hellooThis is 3sha 


In [ ]:
#method 2
import string
def remove_punct(text):
  punct_marks=string.punctuation
  for char in punct_marks:
    text=text.replace(char,'')
  return text
text="helloo!!,.This is 3sha ####"
print(remove_punct(text))

hellooThis is 3sha 


REMOVING HTML TAGS:

In [ ]:
import re
def remove_html_tags(text):
  pattern=re.compile('<.*>') # greedy -it removes all the matched literal from 1st < to lat >
  return pattern.sub(r'',text)
text='<b>Hello</b> <i>World</i>'
print(remove_html_tags(text)) #return empty string

In [ ]:
import re
def remove_html_tags(text):
  pattern=re.compile('<.*?>') # nongreedy -it removes only matched pattern
  return pattern.sub(r'',text)
text='<b>Hello</b> <i>World</i>'
print(remove_html_tags(text)) #removes only html tags and non content

Hello World


REMOVAL URL:

In [ ]:
import re
def remove_url(text):
  pattern=re.compile(r'https?://\S+|www\.\S+')
  return pattern.sub(r'',text)
text="hi this is 3sha https://colab.research.google.com/drive/1dWetXPqeC7OxIfBLbXREFNUoViixYvPf"
print(remove_url(text))

hi this is 3sha 


Problem1:Given a string s, the task is to check if it is Pangram or not.
A pangram is a sentence containing all letters of the English Alphabet.

Examples:

Input: s = “The quick brown fox jumps over the lazy dog”
Output: true
Explanation: The input string contains all characters from ‘a’ to ‘z’.

Input: s = “The quick brown fox jumps over the dog”
Output: false
Explanation: The input string does not contain all characters from ‘a’ to ‘z’, as ‘l’, ‘z’, ‘y’ are missing

In [ ]:
def panagram_check(s):
  s=s.lower()
  alpha_set=set('abcdefghirjklmnopqrstvwxyz')
  s_set=set(s)
  return alpha_set.issubset(s_set)
print(panagram_check("The quick brown fox jumps over the lazy dog"))
print(panagram_check("The quick brwon fox jumps over the dog"))

True
False


In [ ]:
def panagram_check(s):
  check=[False]*26
  s=s.lower()
  for char in s:
    if 'a'<=char<='z':
      index=ord(char)-ord('a')
      check[index]=True
  return all(check)
print(panagram_check("The quick brown fox jumps over the lazy dog"))
print(panagram_check("The quick brwon fox jumps over the dog"))
print(panagram_check("??/.,*7453"))


True
False
False


Problem2:Given a string S that consists of only alphanumeric characters and dashes. The string is separated into n+1 groups by n dashes. We are also given a number K, the task is to reformat the string S, such that each group contains exactly K characters, except for the first group, which could be shorter than K but still must contain at least one character. Furthermore, a dash must be inserted between two groups, and you should convert all lowercase letters to uppercase. Return the reformatted string.

Examples:

Input: S = “5F3Z-2e-9-w”, K = 4
Output: “5F3Z-2E9W”
Explanation: The string S has been split into two parts, each part has 4 characters.Note that two extra dashes are not needed and can be removed.

Input: S = “2-5g-3-J”, K = 2
Output: “2-5G-3J”
Explanation: The string s has been split into three parts, each part has 2 characters except the first part as it could be shorter as mentioned above

In [ ]:
def reformat(s,k):
  s=s.replace('-','').upper()
  res=[]
  while len(s)>k:
    res.append(s[-k:])
    s=s[:-k]
  res.append(s)
  return '-'.join(res[::-1])
s="5F3Z-2e-9-w"
k=4
print(reformat(s,k))
print(reformat("2-5g-3-J",2))

5F3Z-2E9W
2-5G-3J


Problem3:Two strings s1 and s2 are called isomorphic if there is a one-to-one mapping possible for every character of s1 to every character of s2. And all occurrences of every character in ‘s1’ map to the same character in ‘s2’.
Examples:

Input:  s1 = “aab”, s2 = “xxy”
Output: True
Explanation: ‘a’ is mapped to ‘x’ and ‘b’ is mapped to ‘y’.

Input:  s1 = “aab”, s2 = “xyz”
Output: False
Explanation: One occurrence of ‘a’ in s1 has ‘x’ in s2 and other occurrence of ‘a’ has ‘y’.

In [ ]:
def isomorphic_check(s1,s2):
  if len(s1)!=len(s2):
    return False
  mapping={}
  used_chars=set()
  for ch1,ch2 in zip(s1,s2):
    if ch1 in mapping:
      if mapping[ch1]!=ch2:
        return False
    else:
      if ch2 in used_chars:
        return False
      mapping[ch1]=ch2
      used_chars.add(ch2)
  return True
print(isomorphic_check("aab","xxy"))
print(isomorphic_check("aab","xyz"))



True
False


Original:1010230
modify:1111231

In [ ]:

def modify(s):
  n=len(s)
  s=list(s)
  for i in range(n):
    if(s[i]=='0'):
      s[i]='1'
  return "".join(s)
s=input()
print("modify:",modify(s))


102104
modify: 112114


In [ ]:
def modify(s):
  return s.replace('0','1')
s=input()
print(modify(s))


10210304
11211314


find the digit that has to be removed from input1 to convert input1 to a palindrome.

example 1: inpu1:12332 o/p: returns 1
example 2: inp1:251532 o/p: returns 3
example 3: inp1:10101 o/p: returns -1 # as no digit to removed

In [ ]:
def is_palindrome(s):
  return s==s[::-1]
def find_digit_to_remove(n):
  n=str(n)
  left=0
  right=len(n)-1
  while left<right:
    if n[left]!=n[right]:
      if is_palindrome(n[:left]+n[left+1:]):
        return int(n[left])
      elif is_palindrome(n[:right]+n[right+1:]):
        return int(n[right])
      else:
        return "not possible"
    left+=1
    right-=1
  return -1

print(find_digit_to_remove(1010230))
print(find_digit_to_remove(1111231))
print(find_digit_to_remove(251532))
print(find_digit_to_remove(10101))
print(find_digit_to_remove(12332))
print(find_digit_to_remove(23324))



not possible
not possible
3
-1
1
4


INPUT1:"Fi_er"
input 2:"Fever:filer:fixer:filter,fibre:fiber:tailor:offer"
if the input1 and 2 are matched then
output:FILER:FIXER:FIBER
if no match then return "ERROR -009"

In [ ]:
import re
def match_words(pattern,words):
  pattern=pattern.replace('_','.')
  regex=re.compile(pattern,re.IGNORECASE)
  words=words.split(":")
  result=[]
  for word in words:
    if regex.fullmatch(word):
      result.append(word.upper())
  return ":".join(result) if result else "ERROR-009"
pattern="Fi_er"
words="Fever:filer:fixer:filter,fibre:fiber:tailor:offer"
print(match_words(pattern,words))

FILER:FIXER:FIBER


In [ ]:
import re
def match_words(pattern,words):
  pattern=pattern.replace('_','.')
  regex=re.compile(pattern)
  words=words.split(":")
  result=[]
  for word in words:
    if regex.fullmatch(word):
      result.append(word.upper())
  return ":".join(result) if result else "ERROR-009"
pattern="Fi_er"
words="Fever:filer:fixer:filter,fibre:fiber:tailor:offer"
print(match_words(pattern.upper(),words.upper()))


FILER:FIXER:FIBER


In [ ]:
def match_pattern(inp1,inp2):
  res=[]
  words=inp2.split(":")
  for word in words:
    if len(word)!=len(inp1):
      continue
    match=True
    for i in range(len(inp1)):
      if inp1[i]!='_' and inp1[i].lower()!=word[i].lower():
        match=False
        break
    if match:
      res.append(word.upper())
  return ":".join(res) if res else "ERROR-009"
inp1="Fi_er"
inp2="Fever:filer:fixer:filter,fibre:fiber:tailor:offer"
print(match_pattern(inp1,inp2))



FILER:FIXER:FIBER


ADD TWO FRACTIONS:
input1: 2/3
input 2: 5/3
o/p:
22/12
 in mixed farction:1 and 11/12


In [ ]:
def add_fractions(n1,d1,n2,d2):
  lcm=(d1*d2)//gcd(d1,d2)
  num1=n1*(lcm//d1)
  num2=n2*(lcm//d2)
  num=num1+num2
  common=gcd(num,lcm)
  num//=common
  lcm//=common
  return num,lcm
def gcd(a,b):
  while b:
    a,b=b,a%b
  return a
numerator,denominator=add_fractions(2,3,5,4)
print("sum:",numerator,"/",denominator)
if numerator>denominator:
  whole=numerator//denominator
  remainder=numerator%denominator
  print("mixed fraction sum:",whole,"and",remainder,"/",denominator)


sum: 23 / 12
mixed fraction sum: 1 and 11 / 12


WIPRO QUESTIONS:

weight of a string:
a=A=1....Z=z=26

CASE 1:
input1:"Hello world"

input 2:0 (means vowels,special char are ignored)

o/p:89
8+0+12+12+0+0+23+0+18+12+4+0+0=89

CASE 2:
input 1:"Hello world"

input2:1  (means vowels are considered,spcl chars ignored)

o/p:186

0+1212+0+26+0+2+18+0+19+0+0+18+0+0+2+12+0+3+11+0+0+0+23+8+0+20+0=186




In [ ]:
#vowels not considered
def weight_of_str(inp1,inp2):
  vowels=['A','E','I','O','U']
  sum=0
  for i in inp1:
    if i.isalpha():
      upper_char=i.upper()
      if inp2==0 and upper_char in vowels:
        continue
      sum+=ord(upper_char)-ord('A')+1
  return sum
res=weight_of_str("Hello world",0)
print("weight:",res)


weight: 89


In [ ]:
#vowels considered
def weight_of_str(inp1,inp2):
  vowels=['A','E','I','O','U']
  sum=0
  for i in inp1:
    if i.isalpha():
      upper_char=i.upper()
      if inp2==0 and upper_char in vowels:
        continue
      sum+=ord(upper_char)-ord('A')+1
  return sum
res=weight_of_str("Hello world",1)
print("weight:",res)


weight: 124


Findng pin:
i/p="Wipro Technologies"
o/p=8

explanation:
word1 length=5
word2 length=12
tot=5+12=>>17
1+7=8


In [ ]:
def pin(inp1):
  words=inp1.strip().split()
  tot_length=sum(len(word)for word in words)
  while tot_length>=10:
    tot_length=sum(int(digit)for digit in str(tot_length))
  return tot_length
print(pin("Wipro Technologies"))


8


positive and negative number:



In [ ]:
n=int(input("enter num:"))
if n>0:
  print("positive num!")
else:
  print("negative num!")


enter num:-54
negative num!


In [ ]:
n=int(input("enter num:"))
print("positive num!" if n>0 else "negative num!")


enter num:54
positive num!


In [ ]:
def pos_neg(n):
  return "positive num!" if n>0 else "negative num!"
n=int(input("enter input:"))
print(pos_neg(n))

enter input:-54
negative num!


In [ ]:
class pos_neg:
  def __init__(self,n):
    self.n=n
  def check(self):
    return "positive num!" if n>0 else "negative num!"
n=int(input("enter number:"))
res=pos_neg(n)
print(res.check())




enter number:-42
negative num!


In [ ]:
class pos_neg:
  @classmethod
  def check(cls,n):
    if n>0:
      return "positive num!" if n>0 else "negative num!"
n=int(input("enter num:"))
print(pos_neg.check(n))

enter num:43
positive num!


LEAP YEAR


In [ ]:
year=int(input("enter year:"))
if(year%4==0 and year%100!=0 or year%400==0):
  print("leap year!")
else:
  print("not a leap year!")

enter year:1900
not a leap year!


In [ ]:
def leap_year(year):
  return "leap year!" if(year%4==0 and year%100!=0 or year%400==0) else "not a leap year!"
year=int(input("enter year:"))
res=leap_year(year)
print(res)

enter year:1900
not a leap year!


In [ ]:
class leap_year:
  def __init__(self,year):
    self.year=year
  def check(self):
    return "leap year!" if(year%4==0 and year%100!=0 or year%400==0) else "not a leap year!"
year=int(input("enter year:"))
print(leap_year(year).check())

enter year:1900
not a leap year!


In [ ]:
class leap_year:
  @classmethod
  def check(cls,year):
    return "leap year!" if(year%4==0 and year%100!=0 or year%400==0) else "not a leap year!"
year=int(input("enter year:"))
print(leap_year.check(year))

enter year:2008
leap year!


SWAPPING NUMBER

In [ ]:
a,b=int(input("enter n1:")),int(input("enter n2:"))
a,b=b,a
print("after swapping:",a,b)

enter n1:23
enter n2:6
after swapping: 6 23


In [ ]:
def swapping(a,b):
  a,b=b,a
  return a,b
a,b=int(input("enter n1:")),int(input("enter n2:"))
print("after swapping:",swapping(a,b))

enter n1:21
enter n2:85
after swapping: (85, 21)


In [ ]:
a,b=int(input("enter n1:")),int(input("enter n2:"))
temp=a
a=b
b=temp
print("after swapping:",a,b)

enter n1:4
enter n2:21
after swapping: 21 4


In [ ]:
def swapping(a,b):
  a=a-b


SUM OF NATURAL NUMBERS


In [ ]:
n=int(input("enter a limit:"))
sum=0
for i in range(1,n+1):
  sum+=i
print("sum of natural numbers:",sum)

enter a limit:10
sum of natural numbers: 55


In [ ]:
#with formula
n=int(input("enter the limit:"))
sum=(n*(n+1))//2
print("sum of natural num:",sum)

enter the limit:10
sum of natural num: 55


In [ ]:
def sum_natural(n):
  return (n*(n+1))//2
n=int(input("enter a limit:"))
res=sum_natural(n)
print("sum:",res)


enter a limit:10
sum: 55


In [ ]:
class sum_natural:
  def __init__(self,n):
    self.n=n
    self.total=0
  def calculate(self):
    for i in range(1,self.n+1):
      self.total+=i
  def display(self):
    print("sum of natural nums:",self.total)
n=int(input("enter a limit:"))
obj=sum_natural(n)
obj.calculate()
obj.display()

enter a limit:10
sum of natural nums: 55


In [ ]:
class sum_natural:
  @classmethod
  def calculate(cls,n):
    total=0
    for i in range(1,n+1):
      total+=i
    return total
n=int(input("enter a limit:"))
res=sum_natural.calculate(n)
print("sum:",res)



enter a limit:10
sum: 55


COUNT THE VOWELS AND CONSONANTS, digit ,special char count IN  A GIVEN STRING

In [ ]:
def count_vow_cons(text):
  vowels='aeiouAEIOU'
  v_count=0
  c_count=0
  d_count=0
  s_count=0

  for char in text:
    if char.isalpha():
      if char in vowels:
        v_count+=1
      else:
        c_count+=1
    elif char.isdigit():
      d_count+=1
    elif not char.isspace():
      s_count+=1
  return v_count,c_count,d_count,s_count
text=input("enter a text:")
vow,cons,digits,special=count_vow_cons(text)
print(f"Vowels count: {vow}\nConsonants count: {cons}\nDigits count: {digits}\nSpecial characters count: {special}")



enter a text:hii this is 3sha0823e@#$@^^
Vowels count: 6
Consonants count: 7
Digits count: 5
Special characters count: 6


COUNT and PRINT THE VOWELS AND CONSONANTS, digit ,special char count IN A GIVEN STRING

output :count
char are in list format

In [ ]:
def count(text):
  v_count=0
  c_count=0
  d_count=0
  s_count=0
  vow=[]
  cons=[]
  dig=[]
  spcl=[]
  vowels="aeiouAEIOU"
  for char in text:
    if char.isalpha():
      if char in vowels:
        v_count+=1
        vow.append(char)
      else:
        c_count+=1
        cons.append(char)
    elif char.isdigit():
      d_count+=1
      dig.append(char)
    elif not char.isspace():
      s_count+=1
      spcl.append(char)
  return v_count,c_count,d_count,s_count,vow,cons,dig,spcl
text=input("enter a text:")
vow,cons,digits,special,v_char,c_char,d_char,s_char=count(text)
print(f"vowels count:{vow}  vowels:{v_char}")
print(f"consonants count:{vow}  consonats:{c_char}")
print(f"digits count:{vow}  digits:{d_char}")
print(f"speacial char count:{vow}  special char:{s_char}")


enter a text:hii hello rwegb324415 fejkrbgrti%^%$*#%&
vowels count:7  vowels:['i', 'i', 'e', 'o', 'e', 'e', 'i']
consonants count:7  consonats:['h', 'h', 'l', 'l', 'r', 'w', 'g', 'b', 'f', 'j', 'k', 'r', 'b', 'g', 'r', 't']
digits count:7  digits:['3', '2', '4', '4', '1', '5']
speacial char count:7  special char:['%', '^', '%', '$', '*', '#', '%', '&']


In [ ]:
# output is string format
def count(text):
  v_count=0
  c_count=0
  d_count=0
  s_count=0
  vow=[]
  cons=[]
  dig=[]
  spcl=[]
  vowels="aeiouAEIOU"
  for char in text:
    if char.isalpha():
      if char in vowels:
        v_count+=1
        vow.append(char)
      else:
        c_count+=1
        cons.append(char)
    elif char.isdigit():
      d_count+=1
      dig.append(char)
    elif not char.isspace():
      s_count+=1
      spcl.append(char)
  return v_count,c_count,d_count,s_count,vow,cons,dig,spcl
text=input("enter a text:")
vow,cons,digits,special,v_char,c_char,d_char,s_char=count(text)
print(f"vowels count:{vow}  \nvowels:{' '.join(v_char)}")
print(f"consonants count:{vow}  \nconsonats:{' '.join(c_char)}")
print(f"digits count:{vow}  \ndigits:{' '.join(d_char)}")
print(f"speacial char count:{vow}  \nspecial char:{' '.join(s_char)}")


enter a text:iuhbguieb 9743 $%&@^
vowels count:5  
vowels:i u u i e

consonants count:5  
consonats:h b g b
digits count:5  
digits:9 7 4 3
speacial char count:5  
special char:$ % & @ ^


COUNT THE VOWELS AND CONSONANTS, digit ,special char count IN A GIVEN STRING   -list comprehension method

In [ ]:
def count(text):
  vowels="aeiouAEIOU"
  v_count=len([i for i in text if i in vowels])
  c_count=len([i for i in text if i not in vowels and i.isalpha()])
  d_count=len([i for i in text if i.isdigit()])
  s_count=len([i for i in text if not i.isspace() and not i.isalpha() ])
  return v_count,c_count,d_count,s_count
text=input("enter the text:")
vow,cons,dig,spcl=count(text)
print("vowels count:",vow)
print("consonants count:",cons)
print("digits count:",dig)
print("special char count:",spcl)


enter the text:iuhjbj079^&)(*
vowels count: 2
consonants count: 4
digits count: 3
special char count: 8




COUNT  and print the letters THE VOWELS AND CONSONANTS, digit ,special char count IN A GIVEN STRING -list comprehension method

In [ ]:
def count(text):
  vowels="aeiouAEIOU"
  vow=[i for i in text if i in vowels]
  cons=[i for i in text if i not in vowels and i.isalpha()]
  dig=[i for i in text if i.isdigit()]
  special=[i for i in text if not i.isspace() and not i.isalpha() ]
  return vow,cons,dig,special
text=input("enter the text:")
vow,cons,dig,spcl=count(text)
print("vowels count:",len(vow))
print("consonats count:",len(cons))
print("digits count:",len(dig))
print("special count:",len(spcl))
print("vowels :",' '.join(vow))
print("consonants:",' '.join(cons))
print("digits:",' '.join(dig))
print("special char:",' '.join(spcl))


enter the text:jhefubf 986451t  2$#!$#&%^
vowels count: 2
consonats count: 6
digits count: 7
special count: 15
vowels : e u
consonants: j h f b f t
digits: 9 8 6 4 5 1 2
special char: 9 8 6 4 5 1 2 $ # ! $ # & % ^



COUNT THE VOWELS AND CONSONANTS, digit ,special char count IN A GIVEN STRING - using recursion

In [ ]:
#count only vowels and consonants
def count(text,i=0,v=0,c=0):
  vowels="aeiouAEIOU"
  if i==len(text):
    return v,c
  if text[i].isalpha():
    if text[i] in vowels:
      v+=1
    else:
      c+=1
  return count(text,i+1,v,c)
text=input("enter a string:")
v_count,c_count=count(text)
print("vowels count:",v_count)
print("consonants count:",c_count)


enter a string:aaahhhh
vowels count: 3
consonants count: 4


In [ ]:
def count(text,i=0,v=0,c=0,d=0,s=0):
  vowels="aeiouAEIOU"
  if i==len(text):
    return v,c,d,s
  elif text[i].isalpha():
    if text[i] in vowels:
      v+=1
    else:
      c+=1
  elif text[i].isdigit():
    d+=1
  elif not text[i].isspace():
    s+=1
  return count(text,i+1,v,c,d,s)
text=input("enter a string:")
v_count,c_count,d_count,s_count=count(text)
print("vowels count:",v_count)
print("consonants count:",c_count)
print("digits count:",d_count)
print("special char count:",s_count)

enter a string:iejhv 89674 ii ssd #$@$#^
vowels count: 4
consonants count: 6
digits count: 5
special char count: 6


Write a program to count how many times each vowel (a, e, i, o, u) appears in a string.

In [ ]:
def occurence(text):
  vowels="aeiou"
  v_count={vowel:0 for vowel in vowels}
  for i in text.lower():
    if i in v_count:
      v_count[i]+=1
  return v_count
text=input("enter a string:")
res=occurence(text)
for v,count in res.items():
  print(f"{v}:{count}")

enter a string:oijouibhvhg oiwewz
a:0
e:1
i:3
o:3
u:1


Write a program to count how many times each specail char appears in a string.

In [ ]:
def occurence(text):
  spcl="!@#$%^&*()<>?/{}[]~`"
  s_count={s:0 for s in spcl}
  for i in text.lower():
    if i in s_count:
      s_count[i]+=1
  return s_count
text=input("enter a string:")
res=occurence(text)
for s,count in res.items():
  print(f"{s}:{count}")

enter a string:092718@#$%^%&*&))~~!@#$^%&(()#%#@%<><>?{}[]
!:1
@:3
#:4
$:2
%:5
^:2
&:3
*:1
(:2
):3
<:2
>:2
?:1
/:0
{:1
}:1
[:1
]:1
~:2
`:0


Write a program to count how many times each digit appears in a num.

In [ ]:
def occurence(text):
  digits="0123456789"
  d_count={d:0 for d in digits}
  for i in text.lower():
    if i in d_count:
      d_count[i]+=1
  return d_count
text=input("enter a string:")
res=occurence(text)
for d,count in res.items():
  print(f"{d}:{count}")

enter a string:90283078512568364758056138564
0:3
1:2
2:2
3:3
4:2
5:5
6:4
7:2
8:5
9:1


Write a program that removes all vowels from a string and prints the result.

In [ ]:
def remove_vowels(text):
  vowels="aeiouAEIOU"
  return ''.join([char for char in text if char not in vowels])
text=input("enter a string:")
res=remove_vowels(text)
print("after removing vowels:",res)

enter a string:beauty!! iooochh
after removing vowels: bty!! chh


Write a program that replaces all vowels with (1,2,3,4,5) from a string and prints the result.

In [ ]:
def replace_vowels(text):
  vowel_map = {
        'a': '1', 'A': '1',
        'e': '2', 'E': '2',
        'i': '3', 'I': '3',
        'o': '4', 'O': '4',
        'u': '5', 'U': '5'
    }
  return ''.join([vowel_map[char] if char in vowel_map else char for char in text])
text=input("enter a string:")
res=replace_vowels(text)
print("after replacing vowels:",res)

enter a string:hii helllo 3sha
after replacing vowels: h33 h2lll4 3sh1


Take a string input and print only the vowels present in it.

In [ ]:
def vowels(text):
  vowels="aeiouAEIOU"
  return ' '.join([char for char in text if char in vowels])
text=input("enter the string:")
res=vowels(text)
print("vowels in the string:",res)

enter the string:hii hello 3sha
vowels in the string: i i e o a


Take a string input and print only the consonants present in it.

In [ ]:
def consonants(text):
  vowels="aeiouAEIOU"
  return ' '.join([char for char in text if char.isalpha() and char not in vowels])
text=input("enter the string:")
res=consonants(text)
print("consonants in the string:",res)

enter the string:hii hello 3sha
consonants in the string: h h l l s h
